### DISCUSSION TITLE
ConcurrenctionException can not be Catched

### DISCUSSION THREAD

**Discussion by sbgmk** [ravendb/ravendb#17423] [Q&A] [answered]
I insert some Config into my Ravendb and use optimistic concurrency. If i insert the same Document the second time with the same Name the ConcurrencyException is getting Thrown. Thats the intended behaviour. But even if i Catch the ConcurrencyException i got the same Exception with the RavenDB error in the middleware again. This behaviour is unexpected. 
My question is: Where is the second Exception being thrown from and how can i Catch this or prevent it.

We use ASP.Net 6 and the below code gets called in our API Controller.

```
      try
       {
            var workConf= new WorkConf  {  Name = dto.Name  };

            await _session.StoreAsync(workConf);
            await _session.StoreAsync(new Reserved{ ReservedFor = workConf.Id }, "Name/" + workConf.Name);
            await _session.SaveChangesAsync();

            var workConfRef = new WorkConfReference(id: workConf.Id, name: workConf.Name);
            return DomainResult.Success(workConfRef);
        }
        catch (ConcurrencyException e)
        {
            return DomainResult.Conflict<WorkConfReference>(_localizer["Item_Not_Unique"].Value);
        }
```

**Comment by ayende** [comment, marked as answer]
Are you using the _same_ session again? Once you got `ConcurrencyExcpetion` you should create a new session.

If you are calling `SaveChangesAsync` in the middleware, it does the same operation, and it will fail again.

You can call `session.Advanced.Clear()` to reset the entire state, so the middleware call with be a noop.

**Reply by sbgmk** [reply]
Thank you for your Awnser. session.Advanced.Clear() solved the Problem for me.
